src/App/Fossa/LicenseScanner.hs

```diff
@@ -85,11 +87,51 @@ runLicenseScanOnDir ::
   ( Has Diagnostics sig m
   , Has (Lift IO) sig m
   , Has Exec sig m
+  , Has ReadFS sig m
   ) =>
   Text ->
   Path Abs Dir ->
   m [LicenseUnit]
-runLicenseScanOnDir pathPrefix scanDir = withThemisAndIndex $ themisRunner pathPrefix scanDir
+runLicenseScanOnDir pathPrefix scanDir = do
+  -- license scan the root directory
+  rootDirUnits <- withThemisAndIndex $ themisRunner pathPrefix scanDir
+  -- recursively unpack archives and license scan them too
+  otherArchiveUnits <- runFinally $ recursivelyScanArchives pathPrefix scanDir
+  -- when we scan multiple archives, we need to combine the results
+  pure $ combineLicenseUnits (rootDirUnits <> otherArchiveUnits)
+
+recursivelyScanArchives :: (Has (Lift IO) sig m, Has ReadFS sig m, Has Diagnostics sig m, Has Finally sig m, Has Exec sig m) => Text -> Path Abs Dir -> m [LicenseUnit]
+recursivelyScanArchives pathPrefix dir = flip walk' dir $
+  \_ _ files -> do
+    let process file unpackedDir = do
+          let updatedPathPrefix = pathPrefix <> getPathPrefix dir (parent file)
+          currentDirResults <- withThemisAndIndex $ themisRunner updatedPathPrefix unpackedDir
+          recursiveResults <- recursivelyScanArchives updatedPathPrefix unpackedDir
+          pure $ currentDirResults <> recursiveResults
+    archives <- traverse (\file -> withArchive' file (process file)) files
```

Comment:
[nit]: Add a note that `withArchive'` returns `Nothing` if the file is not a valid or supported archive, that's a critical point for how this works and it not obvious here.